homework-tips.md

```diff
@@ -0,0 +1,83 @@
+as follows (the name of the branch depends on the lesson):
+```
+git checkout -b week1
+```
+Then you are ready to do the programming assignment.  Periodically, as you complete parts of the lesson,
+you can add and commit the changes you have made, and you may even push them to your github:
+```
+git add -A
+git commit -m "This commit is for parts 1, 2 and 3 of the lesson"
+git push origin week1
+```
+(The push is optional.)  Providing a descriptive comment is a very good habit to have.
+Once you have finished the lesson, do a git add, commit, and push again.  Now you
+are ready to create a pull request for your lesson. Go to your github in your browser.  Find the repository
+you created with the fork.  Click on pull requests, new pull request.  The base repository for the
+pull request must be the main or master branch of your repository on your github.  Be sure that the
+base is not for the Code-The-Dream-School repository, because if it is, you will not be able to merge
+your branch after the review is complete.  The head repository for the pull request must be your lesson
+branch.  Once you have created the pull request, click on Files changed.  You should see the changes
+you made.  Then click on pull requests again, and copy the link for the pull request you just made.
+This is what you submit using the homework submission form.
+
+Do NOT merge your pull request yet!  Your assignment must be reviewed.  Your reviewer may request changes.
+The reviewer will make comments on the pull request.
+If changes are necessary, you make them to the same lesson branch.  Then you add, commit, and push
+your changes.  The additional changes are automatically added to the same pull request.
+Once your reviewer has approved your lesson pull request, THEN you merge the pull request.
+
+## Creativity and Experimentation
+
+Once you have finished the assignment, you can be creative and make variations on the instructions
+and what the instructor shows you. Through creative experimentation, 
+you will learn much more than you would if you had just
+followed instructions. Of course, there is also a risk of changing so much that you get 
+confused, overwhelmed, and maybe stuck, so for safety, you do your experimentation in a special branch.
+For example, if you have just completed week 1, while the week1 branch is active, you could do:
+```
+git branch -b week1plus
+```
+Then, try some things! Once you are done experimenting, add and commmit your changes.  You don't
+need to push them to github. Again, creativity is crucial for learning. It also allows you to take full advantage of the resources CTD provides. Can't figure something out? Bring your question to a review session! Not sure if you did something the right way? Your assignment reviewer will tell you! 
+
+Subsequent lessons build on the previous ones, so you should checkout
+the week1 branch before creating the lesson branch for week2.
+
+## Recovering From a Common Mistake
```

Comment:
Another common thing I've seen: 

*   student makes changes in week1 branch and has it up for review
*   student checks out a new branch and starts working on week2 assignment
*   student gets feedback on week1 branch and adds some changes 

Sometimes the student ends up adding the new changes to `week2` branch instead of `week1`. Other times, they add them to `week1` but forget to pull those changes into `week2` so they end up with merge conflicts or accidentally reverting some changes when they put their `week2` PR up for review.

Reply:
Ok, I have added some text about that -- however, there are cases where they will get merge conflicts when they try to merge week1 changes into week2.  If they do, I am saying they should ask a mentor for help.